src/lib/printer.ts

```diff
@@ -0,0 +1,172 @@
+        ]
+      },
+      copies: {
+        default_copies: 1,
+        max_copies: 100
+      },
+      media_size: {
+        option: [
+          {
+            name: "ISO_A4",
+            width_microns: 210000,
+            height_microns: 297000,
+            is_default: true
+          },
+          {
+            name: "NA_LETTER",
+            width_microns: 215900,
+            height_microns: 279400
+          }
+        ]
+      },
+      page_orientation: {
+        option: [
+          {
+            type: "PORTRAIT",
+            is_default: true
+          },
+          { type: "LANDSCAPE" },
+          { type: "AUTO" }
+        ]
+      },
+      duplex: {
+        option: [
+          { type: "NO_DUPLEX", is_default: true },
+          { type: "LONG_EDGE" },
+          { type: "SHORT_EDGE" }
+        ]
+      }
+    }
+  });
+}
+
+/**
+ * Printer capabilities request callback type
+ */
+type PrinterCapabilitiesCallback = (p: unknown) => void;
+
+/**
+ * Returns a list of "virtual" printers,
+ * in our case "Print/Publish to Arweave"
+ */
+export function getPrinters(callback: PrinterInfoCallback) {
+  callback([
+    {
+      id: ARCONNECT_PRINTER_ID,
+      name: "Print to Arweave",
+      description:
+        "Publish the content you want to print on Arweave, permanently."
+    }
+  ]);
+}
+
+/**
+ * Printer info request callback type
+ */
+type PrinterInfoCallback = (p: chrome.printerProvider.PrinterInfo[]) => void;
+
+/**
+ * Handles the request from the user to print the page to Arweave
+ */
+export async function handlePrintRequest(
+  printJob: chrome.printerProvider.PrintJob,
+  resultCallback: PrintCallback
+) {
+  // only print for the ArConnect printer
+  if (printJob.printerId !== ARCONNECT_PRINTER_ID) return;
+
+  // wallet
+  let decryptedWallet: DecryptedWallet;
+
+  try {
+    // build data blog
+    const data = new Blob([printJob.document], { type: printJob.contentType });
+
+    // get user wallet
+    decryptedWallet = await getActiveKeyfile();
+
+    if (decryptedWallet.type === "hardware")
+      throw new Error("Cannot print with a hardware wallet.");
+
+    // extension manifest
+    const manifest = browser.runtime.getManifest();
+
+    // setup tags
+    const tags = [
+      { name: "App-Name", value: manifest.name },
+      { name: "App-Version", value: manifest.version },
+      { name: "Type", value: "Archive" },
+      { name: "Content-Type", value: printJob.contentType },
+      { name: "print:title", value: printJob.title },
+      { name: "print:timestamp", value: new Date().getTime().toString() }
+    ];
+
+    // create data item
+    const dataSigner = new ArweaveSigner(decryptedWallet.keyfile);
+    const dataEntry = createData(
+      new Uint8Array(await data.arrayBuffer()),
+      dataSigner,
+      { tags }
+    );
+
+    // sign an upload data
+    await dataEntry.sign(dataSigner);
+    await uploadDataToTurbo(dataEntry, "https://turbo.ardrive.io");
```

Comment:
@martonlederer Since Turbo has a limit of ~100KB for free uploads, and additional credits/balance are required for larger files, we currently can't upload files larger than 100KB using Turbo alone. Should we consider falling back to direct Arweave uploads if Turbo upload fails ? WDYT?

Reply:
I'm honestly not sure. I think bundled txes should be preferred, but what you're saying makes complete sense.

Reply:
Yes bundled txes should be preferred, but `Print to Arweave` may not work for most cases, as the data size will likely exceed the 100KB limit and users mayn't have Turbo credits. The only pages I could print were simple pages with minimal content otherwise it failed to upload. (**Note**: I don't have Turbo credits/balance.)

Reply:
I agree @pawanpaudel93, let's fall back to Arweave when Turbo upload fails